src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -50,6 +51,24 @@ public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> so
         private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource> source, int size)
         {
             using IEnumerator<TSource> e = source.GetEnumerator();
+
+            // TODO what should the threshold be?
```

Comment:
Might be able to determine that using benchmarks. After what size does `Array.Resize` become more costly than using a resizable builder?

Reply:
> After what size does Array.Resize become more costly than using a resizable builder?

The worst case for the `Array.Resize` approach is when we have just one element; in that case we have to allocate an array of `Size` when we could have allocated barely anything with `List<T>`. If we use that metric we'd end up with a very low threshold for `List<T>` being better (~50 on my machine).

The best case for the array approach is when we have exactly `Size` elements and don't need to resize at all. That should win out over `List<T>` for any size.

Without making assumptions about chunk size and the number of elements, there's no perfect threshold. That said, any reasonably small threshold avoids the worst case for the array approach.